draft-ietf-quic-transport.md

```diff
@@ -941,20 +941,15 @@ connection ID for the duration of the connection or until its peer invalidates
 the connection ID via a RETIRE_CONNECTION_ID frame
 ({{frame-retire-connection-id}}).
 
-Endpoints store received connection IDs for future use.  An endpoint that
-receives excessive connection IDs MAY discard those it cannot store without
-sending a RETIRE_CONNECTION_ID frame.  An endpoint that issues connection IDs
-cannot expect its peer to store and use all issued connection IDs.
-
 An endpoint SHOULD ensure that its peer has a sufficient number of available and
-unused connection IDs.  While each endpoint independently chooses how many
-connection IDs to issue, endpoints SHOULD provide and maintain at least eight
-connection IDs.  The endpoint SHOULD do this by always supplying a new
-connection ID when a connection ID is retired by its peer or when the endpoint
-receives a packet with a previously unused connection ID.  Endpoints that
-initiate migration and require non-zero-length connection IDs SHOULD provide
-their peers with new connection IDs before migration, or risk the peer closing
-the connection.
+unused connection IDs. Endpoints store received connection IDs for future use.
+An endpoint uses a transport parameter to advertise the number of connection IDs
+it can store for future use. An endpoint SHOULD NOT provide more connection IDs
+than this limit. If an endpoint has provided its peer with the maximum number of
+connection IDs, it SHOULD only provide a new connection ID when the peer retires
+one. Endpoints that initiate migration and require non-zero-length connection
+IDs SHOULD provide their peers with new connection IDs before migration, or risk
+the peer closing the connection.
```

Comment:
There are a lot of SHOULDs here, but it's a bit light on the associated justification.  I think that we'll need some more text about why an endpoint might choose not to follow this advice.  Having everything in the one paragraph will make that difficult.